src/java/org/apache/cassandra/db/compaction/UnifiedCompactionContainer.java

```diff
@@ -65,6 +66,7 @@ public class UnifiedCompactionContainer implements CompactionStrategyContainer
         this.strategy = new UnifiedCompactionStrategy(factory, backgroundCompactions, params.options());
         this.enabled = new AtomicBoolean(enabled);
         this.enableAutoCompaction = enableAutoCompaction;
+        this.hasVector = factory.getRealm().metadata().hasVectorType();
```

Comment:
`factory.getRealm().metadata()` can change concurrently with the construction.

For a bit of extra safety I would derive the value of `hasVector` from the metadata instance that `Controller.fromOptions` used to construct it. I guess the easiest way to do this is to store and use a field in `Controller`.